src/transformers/modeling_tf_utils.py

```diff
@@ -1157,6 +1157,8 @@ def build(self, input_shape=None):
             self.built = True
         else:
             self.built = True
+            # Set the serving spec quickly to ensure that Keras doesn't use the specific dummy input shapes as the spec
+            self._set_save_spec(self._prune_signature(self.input_signature))
```

Comment:
Maybe explain a bit why we don't do this in init method?

Reply:
It's a bit of a long story! `_set_save_spec` is normally called internally by Keras, and for subclassed models (i.e. all models in `transformers`) it uses **the first input shapes the model sees.** This was a huge problem for us, because we'd pass in some tiny dummy inputs and it would just lock in that useless specific shape as the model's save spec. This made exporting/serving a real nightmare!

We avoid that by setting a correct, general save spec before the model has seen any inputs. It doesn't really matter in most cases whether we put that in the `__init__` or the `build` method, as long as it happens before we pass dummy inputs in. However, there is one edge case where it makes a small difference: If the user builds a model from a config, and then passes inputs of a specific shape in. In this case, putting it in `build()` allows the user to set the save spec with their own inputs, which can be useful in a couple of cases.

I'm not convinced this is a perfect solution, but it resolves an edge case in our in-graph tokenizer test, so it seems a little better than the alternative!

Reply:
So a short version could be 
```python
# put this in build() allows the user to set the save spec with their own inputs.
```
😄 ...?

Reply:
Sure! I'll comment something like that